instrumentation/src/javatests/org/oppia/android/instrumentation/topic/TopicTest.kt

```diff
@@ -0,0 +1,84 @@
+package org.oppia.android.instrumentation.topic
+
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.uiautomator.UiDevice
+import com.google.common.truth.Truth.assertThat
+import org.junit.Before
+import org.junit.Test
+import org.oppia.android.instrumentation.testing.EndToEndTestHelper.findObjectByRes
+import org.oppia.android.instrumentation.testing.EndToEndTestHelper.findObjectByText
+import org.oppia.android.instrumentation.testing.EndToEndTestHelper.scrollRecyclerViewTextIntoView
+import org.oppia.android.instrumentation.testing.EndToEndTestHelper.startOppiaFromScratch
+import org.oppia.android.instrumentation.testing.EndToEndTestHelper.waitForRes
+import org.oppia.android.instrumentation.testing.EndToEndTestHelper.waitForText
+
+/** Tests for TopicActivity. */
+class TopicTest {
+  private lateinit var device: UiDevice
+
+  @Before
+  fun setUp() {
+    // Initialize UiDevice instance
+    device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+    device.startOppiaFromScratch()
+  }
+
+  @Test
+  fun testTopicActivity_exploredFully_finishedSuccessfully() {
```

Comment:
It feels like each of these distinct blocks ought to be independent tests since they're verifying different aspects of the topic experience. This is a bit different than the exploration test since that needs to play through a whole exploration to verify that it can.

I think this might be slightly in contradiction with earlier feedback from @FareesHussain (Farees: please follow up if you disagree with this).

Reply:
We had distinct blocks in exploration as well, each interaction had a different aspect. I believe it should be same here as well, each tab or component are independent. Only reason I can think of for having multiple tests is when we have interdependent topics.

I can't think of a reason for having independent tests.

Reply:
Generally, we should have 1 test method per behavior that we're testing. However, this doesn't translate as cleanly for e2e tests since the covered behaviors are so high-level. Instead, I think we should define them as 1 test per single user flow that's being verified. In the case of explorations, we had 1 test to verify that we could play through and finish an exploration. Any other scenarios (such as verifying that hints work in various cases) would need to be in a separate test method rather than being added onto the existing test method.

It seems to me that this test, unlike the one for explorations, is actually verifying multiple scenarios and therefore would benefit from the cleanliness of having distinct test methods.

Note that I'm also thinking about how these suites are going to evolve over time (one possibility is that many of our QA tests may end up being built as UiAutomator tests so establishing a clear standard here is important).

WDYT?

Reply:
Sounds good, also had a discussion with Vinita and sean regarding this
@ayush0402 PTAL